opencog/nlp/sureal/SuRealPMCB.cc

```diff
@@ -97,212 +86,268 @@ bool SuRealPMCB::variable_match(const Handle &hPat, const Handle &hSoln)
+    if (qAllPatNodes.size() != qAllSolnNodes.size())
+        return false;
 
-    // helper for sorting EvaluationLinks in word sequence order
-    auto sortRightInsts = [](const Handle& h1, const Handle& h2)
+    // compare the disjuncts for each of the nodes
+    for (size_t i = 0; i < qAllPatNodes.size(); i++)
     {
-        // get the ListLinks from the EvaluationLinks
-        const Handle& hListLink1 = LinkCast(h1)->getOutgoingSet()[1];
-        const Handle& hListLink2 = LinkCast(h2)->getOutgoingSet()[1];
+        Handle& hPatNode = qAllPatNodes[i];
+        Handle& hSolnNode = qAllSolnNodes[i];
 
-        // get the second WordInstanceNodes from the ListLinks
-        const Handle& hWordInst1 = LinkCast(hListLink1)->getOutgoingSet()[1];
-        const Handle& hWordInst2 = LinkCast(hListLink2)->getOutgoingSet()[1];
+        // move on if the pattern node is a VariableNode or an InperpretationNode
+        if (hPatNode->getType() == VARIABLE_NODE || hPatNode->getType() == INTERPRETATION_NODE)
+            continue;
 
-        // get the NumberNodes from the WordSequenceLinks
-        Handle hNumNode1 = get_neighbors(hWordInst1, false, true, WORD_SEQUENCE_LINK)[0];
-        Handle hNumNode2 = get_neighbors(hWordInst2, false, true, WORD_SEQUENCE_LINK)[0];
+        // get the corresponding WordNode of the pattern node
+        Handle hPatWordNode;
+        auto it = m_words.find(hPatNode);
+        if (it == m_words.end())
+        {
+            std::string sPat = m_as->get_name(hPatNode);
+            std::string sPatWord = sPat.substr(0, sPat.find_first_of('@'));
 
-        // compare their word sequences
-        return NodeCast(hNumNode1)->getName() < NodeCast(hNumNode2)->getName();
-    };
+            // get the WordNode associated with the word (extracted from "word@1234" convention)
+            hPatWordNode = m_as->get_handle(WORD_NODE, sPatWord);
 
-    // sort the qLGInstsLeft in reverse word sequence order
-    std::sort(qLGInstsLeft.begin(), qLGInstsLeft.end(), sortLeftInsts);
+            m_words.insert({hPatNode, hPatWordNode});
+        }
+        else hPatWordNode = it->second;
 
-    // sort the qLGInstsRight in word sequence order
-    std::sort(qLGInstsRight.begin(), qLGInstsRight.end(), sortRightInsts);
+        // get the corresponding WordInstanceNode of the solution node
+        std::string sSoln = m_as->get_name(hSolnNode);
+        Handle hSolnWordInst = m_as->get_handle(WORD_INSTANCE_NODE, sSoln);
 
-    // get the LG connectors for those in the qLGInstsLeft
-    for (Handle& hEvalLink : qLGInstsLeft)
-    {
-        const Handle& hLinkInstNode = LinkCast(hEvalLink)->getOutgoingSet()[0];
+        // sometimes when the pattern node is not a variable, it may match to
+        // a solution node that does not have a corresponding WordInstanceNode,
+        // so we should accept it here, e.g. (PredicateNode "possession")
+        if (hSolnWordInst == Handle::UNDEFINED)
```

Comment:
Almost thought this is wrong.  Can you add an explanation saying something like "if the node is a variable, it would have matched to a node with corresponding WordInstanceNode by variable match", and "if the node is not a variable, then the standard node_match would have made it matched to itself".